src/analysis/dom/domtree.rs

```diff
@@ -1,443 +1,159 @@
-impl DFSVisitor {
-    pub fn new() -> DFSVisitor {
-        DFSVisitor {
-            post_order: Vec::new(),
-            pre_order: Vec::new(),
-            visited: Vec::new(),
-        }
-    }
-
-    pub fn dfs<N, E>(&mut self, g: &Graph<N, E>, node: graph::NodeIndex) {
-        if self.visited.contains(&node) {
-            return;
-        }
-
-        self.pre_order.push(node);
-        let direction = EdgeDirection::Outgoing;
-        let neighbors_iter = g.neighbors_directed(node, direction)
-                              .collect::<Vec<graph::NodeIndex>>();
-
-        for n in &neighbors_iter {
-            self.dfs(g, *n);
-        }
-
-        self.visited.push(node);
-        self.post_order.push(node);
-    }
+use petgraph::{graph, EdgeDirection};
+use petgraph::algo::dominators;
+use petgraph::algo::dominators::{Dominators, DominatorsIter};
+use petgraph::graph::Graph;
+use petgraph::graph::NodeIndex;
 
-    pub fn post_order(&self) -> Vec<graph::NodeIndex> {
-        self.post_order.clone()
-    }
+use petgraph::visit::{Dfs, Bfs, Walker};
+use std::collections::{HashMap, HashSet};
+use std::collections::hash_map::Iter;
+use std::default;
 
-    pub fn pre_order(&self) -> Vec<graph::NodeIndex> {
-        self.pre_order.clone()
-    }
-}
+// TODO: make the functoins work on domintor tree / post dominator tree by passing them
+// an enum i.e call dominators(n, Domination:Pre), dominators(n, Domination:Post) to
+// find the dominators / post dominators of a given node
 
+/// Contains dominance info for a given graph
+/// If you want to find a NodeIndex from a usize value, use the node_map.first_to_second
 #[derive(Clone, Debug)]
-/// Wrapper struct that consolidates the `Dom` and `PostDom` information for a
-/// Graph.
-pub struct DomInfo {
-    dom: Option<DomTree>,
-    postdom: Option<DomTree>,
-}
-
-impl default::Default for DomInfo {
-    fn default() -> DomInfo {
-        DomInfo {
-            dom: None,
-            postdom: None,
-        }
-    }
+struct DomminanceInfo<N, E> {
+    dominators: Dominators<NodeIndex>,
+
+    post_dominators: Dominators<NodeIndex>,
+    // Dominance frontier for each node
+    frontier_map: HashMap<NodeIndex, HashSet<NodeIndex>>,
+    // Predecessors of each node
+    preds_map: HashMap<NodeIndex, Vec<NodeIndex>>,
+    // Needed to acess elements by dfs order in an O(1) operation
+    node_map: BiMap<usize, NodeIndex>,
+    // Graph
+    g: Option<graph::Graph<N, E>>,
 }
 
-impl DomInfo {
-    pub fn new() -> DomInfo {
-        DomInfo {
-            dom: None,
-            postdom: None,
-        }
-    }
-
-    /// Constructs the DomTree and computes the dominance frontier for `g`.
-    pub fn build_dom_tree<N, E>(&mut self, g: &Graph<N, E>, i: graph::NodeIndex)
-        where N: Clone,
-              E: Clone
+impl<N, E> DomminanceInfo<N, E> {
+    /// Computes info about dominance relation in a given graph
+    fn new(g: &Graph<N, E>, start_node: graph::NodeIndex) -> DomminanceInfo<N, E>
```

Comment:
I replaced the `Graph` with `StableDiGraph<NodeData,EdgeData>` just for now to avoid the fuss of generic traits ( as I'm still new to Rust ), I'll focus on that after it works first.

The `StableDiGraph` doesn't have a reverse function unlike the one of `Graph`, 
should I implement it ? am I going in the right way ?